packages/product/src/services/product-module-service.ts

```diff
@@ -405,6 +398,229 @@ export default class ProductModuleService<
+          })
+        )
+
+        const products = await this.productService_.update(
+          productsData,
+          sharedContext
+        )
+
+        const productByIdMap = new Map<string, TProduct>(
+          products.map((product) => [product.id, product])
+        )
+
+        const productOptionsData = [...productOptionsMap]
+          .map(([id, options]) => options.map((option) => ({
+            ...option,
+            product: productByIdMap.get(id)!,
+          })))
+          .flat()
+
+        const productOptions = await this.productOptionService_.create(
+          productOptionsData,
+          sharedContext
+        )
+
+        let productVariantIdsToDelete: string[] = []
+        const productVariantsToCreateMap = new Map<
+          string,
+          ProductTypes.CreateProductVariantDTO[]
+        >()
+
+        const productVariantsToUpdateMap = new Map<
+          string,
+          ProductTypes.UpdateProductVariantDTO[]
+        >()
+
+        for (const [productId, variants] of productVariantsMap) {
+          const variantsToCreate: ProductTypes.CreateProductVariantDTO[] = []
+          const variantsToUpdate: ProductTypes.UpdateProductVariantDTO[] = []
+          const existingVariants = existingProductVariantsMap.get(productId)
+
+          variants.forEach((variant) => {
+            const isVariantIdDefined = ("id" in variant) && isDefined(variant.id)
+            const existingProductVariant = existingVariants
+              ?.find((existingVariant) => isVariantIdDefined && (existingVariant.id === variant.id))
+
+            if (isVariantIdDefined && !existingProductVariant) {
+              throw new MedusaError(
+                MedusaError.Types.NOT_FOUND,
+                `ProductVariant "${variant.id}" not found for product "${productId}"`
+              )
+            }
+
+            if (isVariantIdDefined) {
+              variantsToUpdate.push(variant as ProductTypes.UpdateProductVariantDTO)
+            } else {
+              variantsToCreate.push(variant as ProductTypes.CreateProductVariantDTO)
+            }
+
+            const variantOptions = variant.options?.map((option, index) => {
+              const productOption = productOptions[index]
+              return {
+                option: productOption,
+                value: option.value,
+              }
+            })
+
+            if (variantOptions) {
+              variant.options = variantOptions
+            }
+          })
+
+          productVariantsToCreateMap.set(productId, variantsToCreate)
+          productVariantsToUpdateMap.set(productId, variantsToUpdate)
+
+          const variantsToUpdateIds = variantsToUpdate.map(v => v?.id) as string[]
+          const existingVariantIds = existingVariants?.map(v => v.id) || []
+          const variantsToUpdateSet = new Set(variantsToUpdateIds)
+
+          productVariantIdsToDelete = productVariantIdsToDelete.concat([
```

Comment:
todo: `productVariantIdsToDelete.push(...Set)`

Reply:
This will push an array inside the array. I'd rather work with one dimension at every point.

Reply:
no it will pus all elements in the array not an array in the array right?

Reply:
You're right. Changed it! 🎉